### ISSUE TITLE
[🚀 Feature]: switch Java-tests from JUnit 4 to JUnit 5

### ISSUE THREAD

**Issue by christophs78** [SeleniumHQ/selenium#10196]
### Feature and motivation

Selenium still uses JUnit 4 (see eg https://github.com/SeleniumHQ/selenium/blob/702b3aae7385adf331d75cdeacd79720cf6b9c84/java/maven_deps.bzl#L56, https://github.com/SeleniumHQ/selenium/blob/702b3aae7385adf331d75cdeacd79720cf6b9c84/java/test/org/openqa/selenium/testing/JUnit4TestBase.java#L35)
JUnit 5 is out since 2017 and widely adopted.

I´m asking for this because there are issues when using JUnit 5 and Selenium 4 together. (https://github.com/SeleniumHQ/selenium/issues/9359) IMO such issues already should be recognized during automated testing.

**Comment by titusfortner** [comment]
We use bazel not maven to run tests in parallel, and we test with Java 11, either one of which would keep issue #9359 from being surfaced. We also don't test TestNG, etc. I don't think catching that error is either the fault of or the responsibility of the Selenium test suite.

I don't there is a compelling reason to have an open issue to update our code base, though we might be open to Pull Requests for it. I'll see what others think.

**Comment by titusfortner** [comment]
So, yes, this is a good idea, but it would be a major undertaking and lower on our priority list. We'll leave it open; anyone who wants to make a Pull Request to help with this, it would be appreciated.

**Comment by christophs78** [comment]
I may look into this within the next weeks or months. My offer would be to work on this in batches. e.g. move some packages from junit4 to junit5, do a PR. after this come back, do some more packages, open another PR.
Does this make sense to you?

**Comment by diemol** [comment]
Thank you, it does make sense.

**Comment by bonigarcia** [comment]
The first step for migrating the Selenium tests from JUnit 4 to JUnit 5 has been implemented in #10752. This PR (already merged) enables the Bazel [contrib_rules_jvm](https://github.com/bazel-contrib/rules_jvm), which allows using the [JUnit Platform](https://junit.org/junit5/docs/current/user-guide/) for running test cases. The macro used to execute test suites in Selenium is called `java_test_suite`, and so far uses the JUnit 4 runner by default since most existing tests are JUnit 4-based.

In summary, currently, the existing Selenium tests are already executed on the JUnit Platform (i.e., JUnit 5), using the JUnit Vintage engine (i.e., JUnit 4) by default.

The next step will be migrating the JUnit 4 to JUnit Jupiter (i.e., the new programming model supported by JUnit 5). To gradually change this in the Selenium test codebase, I think the most suitable approach is to migrate first the JUnit 4 test base to JUnit Jupiter. This change means reimplementing the following foundation classes as a [JUnit 5 extension](https://junit.org/junit5/docs/current/user-guide/#extensions): [JUnit4TestBase](https://github.com/seleniumHQ/selenium/blob/trunk/java/test/org/openqa/selenium/testing/JUnit4TestBase.java), [SeleniumTestRunner](https://github.com/seleniumHQ/selenium/blob/trunk/java/test/org/openqa/selenium/testing/SeleniumTestRunner.java), and [SeleniumTestRule](https://github.com/seleniumHQ/selenium/blob/trunk/java/test/org/openqa/selenium/testing/SeleniumTestRule.java). Then, all the tests using the `JUnit4TestBase` should use the new JUnit 5 extension. These changes need to be in the same PR.

As a final step, we can migrate the rest of the existing JUnit 4 tests to JUnit Jupiter. This change involves using the new programming model in all tests (e.g., changing `org.junit.Test` to `org.junit.jupiter.api.Test`, `org.junit.Before`  to `org.junit.jupiter.api.BeforeEach`, assertions, assumptions, parameterized tests, etc.). This change (or changes) can be a separate PR (or several).

I can do the PR to migrate the JUnit 4 scaffolding classes (test base, runner, and rules) to JUnit 5.

**Comment by zzhilin** [comment]
Hi, I'm a participant in Codeday Labs and my project team was going to contribute to this specific issue. Now that it's already closed, I'm wondering if there's any additional work to be done, specifically the final step mentioned above to migrate rest of the tests to JUnit Jupiter. Thank you!

**Comment by diemol** [comment]
Apologies for not updating the issue. We started working on it and the idea was to migrate a few tests and then let the rest come through contributions. But while doing the work, we noticed that the approach demanded considerable amount of work, so we took the all or nothing path, which means that we did the whole migration and now this work is completed.

Please have a look at other open issues in case you are interested to contribute, either by triaging or fixing bug. We are always available through the chat if questions arise https://www.selenium.dev/support/
